letsencrypt/plugins/common.py

```diff
@@ -157,19 +159,33 @@ def get_cert_file(self, achall):
         return os.path.join(
             self.configurator.config.work_dir, achall.nonce_domain + ".crt")
 
+    def get_key_path(self, achall):
+        """Get standardized path to challenge key."""
+        return os.path.join(
+            self.configurator.config.work_dir, achall.nonce_domain + '.pem')
+
     def _setup_challenge_cert(self, achall, s=None):
         # pylint: disable=invalid-name
         """Generate and write out challenge certificate."""
-        cert_path = self.get_cert_file(achall)
+        cert_path = self.get_cert_path(achall)
         # Register the path before you write out the file
         self.configurator.reverter.register_file_creation(True, cert_path)
 
         cert_pem, response = achall.gen_cert_and_response(s)
 
         # Write out challenge cert
-        with open(cert_path, "w") as cert_chall_fd:
+        with open(cert_path, "wb") as cert_chall_fd:
             cert_chall_fd.write(cert_pem)
 
+        key_path = self.get_key_path(achall)
+        key_pem = achall.key.key.private_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PrivateFormat.TraditionalOpenSSL,
+            encryption_algorithm=serialization.NoEncryption())
+        with le_util.safe_open(key_path, 'wb', chmod=0o400) as key_file:
+            key_file.write(key_pem)
+        self.configurator.reverter.register_file_creation(True, key_path)
```

Comment:
In case of a crash... it doesn't hurt if the reference has been created in the reverter... but it does hurt if the file was created and there is no reference to clean it up.